sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java

```diff
@@ -1737,23 +1746,23 @@ public void testNullStringEquality() throws Exception
         "SELECT COUNT(*)\n"
         + "FROM druid.foo\n"
         + "WHERE NULLIF(dim2, 'a') = null",
+        NullHandling.replaceWithDefault() ?
         ImmutableList.of(
             Druids.newTimeseriesQueryBuilder()
                   .dataSource(CalciteTests.DATASOURCE1)
                   .intervals(querySegmentSpec(Filtration.eternity()))
                   .granularity(Granularities.ALL)
-                  .filters(expressionFilter("case_searched((\"dim2\" == 'a'),"
-                                             + (NullHandling.replaceWithDefault() ? "1" : "0")
-                                             + ",(\"dim2\" == null))"))
+                  .filters(selector("dim2", "a", null))
```

Comment:
Here we have `NULLIF(dim2, 'a') = NULL` being translated as,

- Old, when NullHandling.replaceWithDefault() is on: `CASE WHEN dim2 = 'a' THEN TRUE ELSE dim2 IS NULL END` - equivalent to `dim2 = 'a' OR dim2 IS NULL`
- Old, when NullHandling.replaceWithDefault() is off: `FALSE`
- New, when NullHandling.replaceWithDefault() is on: `dim2 = 'a'`
- New, when NullHandling.replaceWithDefault() is off: `FALSE`

I don't think the new `dim2 = 'a'` translation is really defensible: if it considers a null return from NULLIF due to `dim2 = 'a'` equal to null, then it should also count a null return due to `dim2 IS NULL` (or empty string, which should be equivalent if replaceWithDefault is on). The old translations are both correct, I think, in their respective modes.

I just tried a similar query in PostgreSQL 9.6 and got zero back, which makes sense since PostgreSQL behaves in "NullHandling.replaceWithDefault() is off" (the SQL standard) mode, and should be treating this filter as "always false":

```
SELECT COUNT(*) FROM test WHERE NULLIF(NULL, 'a') = NULL OR NULLIF('a', 'a') = NULL
```

I looked into this for a few minutes but didn't see a reason yet why the translation changed. @xueyumusic - do you have any idea?

Reply:
Ah, I see you pointed out some plan diffs in the PR description:

> 1.17.0
>
> ```
> rel#27:LogicalFilter.NONE.[[]](input=rel#20:Subset#1.NONE.[],condition=CASE(=($0, 'a'), CAST(true):BOOLEAN, =($0, null)))
>   call#64 rule [ReduceExpressionsRule(Filter)]
>     rel#21:LogicalFilter.NONE.[](input=rel#20:Subset#1.NONE.[],condition==(CASE(=($0, 'a'), null, $0), null))
> ```
>
> 1.18.0
>
> ```
> rel#27:LogicalFilter.NONE.[](input=rel#20:Subset#1.NONE.[],condition=IS TRUE(=($0, 'a')))
>   call#64 rule [ReduceExpressionsRule(Filter)]
>     rel#21:LogicalFilter.NONE.[](input=rel#20:Subset#1.NONE.[],condition==(CASE(=($0, 'a'), null, $0), null))
> ```

This appears to be happening in `RexSimplify.simplifyCase`. It doesn't really look right to me, for the reason mentioned above (if it thinks `null = null` due to dim2 being 'a', it should also think that `null = null` due to dim2 being `null`). I believe this functionality was altered in https://issues.apache.org/jira/browse/CALCITE-1413 - will have to dig more into this later. If anyone else finds anything please do post it.